### ISSUE TITLE
Potential Use After Free

### ISSUE THREAD

**Issue by wurongxin1987** [tmux/tmux#1553]
Please read https://raw.githubusercontent.com/tmux/tmux/master/CONTRIBUTING

In the function control_notify_input in the source file notify.c, it will call **evbuffer_pullup** and potentially input->first may be free inside **evbuffer_pullup**. 

![image](https://user-images.githubusercontent.com/4220002/48978359-88e9c980-f0e5-11e8-9a9b-87181e78c4aa.png)

In the function **evbuffer_pullup**,  a local variable **chain** is assigned with the value **buf->first** (buf is the first parameter of the function). Then, it can be freed in certain condition (Please see Line 1428 in buffer.c in the libevent-2.1-6, and it will invoke **evbuffer_chain_free** which will free **chain**.  ). Please see the following figures. 

![image](https://user-images.githubusercontent.com/4220002/48978371-ae76d300-f0e5-11e8-936c-41391e8bac2e.png)

![image](https://user-images.githubusercontent.com/4220002/48978389-e9790680-f0e5-11e8-884c-5ab32b9d2d0e.png)

![image](https://user-images.githubusercontent.com/4220002/48978437-51c7e800-f0e6-11e8-978e-bd1641c61204.png)

If input->first is free, when the next time **notify_input** function calls **control_notify_input**, it will lead to use after free. 

![image](https://user-images.githubusercontent.com/4220002/48978450-9a7fa100-f0e6-11e8-8efc-73288912a1ea.png)

Thanks for your attention. Please have a check whether this is a potentially true bug. Thanks again.

**Comment by wurongxin1987** [comment]
> If this is a problem, it is in libevent not tmux, but I don't think it is because if there isn't enough space and buf->first would be freed, it is replaced a few lines above the loop.

Actually, this use after free can only be triggered by invoking control_notify_input at least twice. I think it is not a problem in libevent since a single call to this function will not lead to any problem. Moreover, buf->first may be freed at the first iteration of the for loop, and may not be replaced a few lines above the loop (considering that it may go to branch at Line 1403).  

![image](https://user-images.githubusercontent.com/4220002/49010626-cebe9480-f1ae-11e8-988d-fc4f8a6c2148.png)
![image](https://user-images.githubusercontent.com/4220002/49010640-db42ed00-f1ae-11e8-9de3-c7445e31f4bb.png)
![image](https://user-images.githubusercontent.com/4220002/49010660-e433be80-f1ae-11e8-9b0b-4a8f6d39a9ce.png)

**Comment by nicm** [comment]
No, tmux does not ask libevent to free this evbuffer, it should be safe to call evbuffer_pullup as many times as needed, if not it is a libevent problem.

**Comment by nicm** [comment]
What you are saying is that this is unsafe:
~~~~
evbuffer_pullup(b, -1);
evbuffer_pullup(b, -1);
~~~~
But it is meant to be safe (assuming not out of memory). If not, then this needs to be fixed in libevent, you will need to report it to them. Try @nmathewson.
